### PR TITLE
Matcherfix

### PR DESCRIPTION
**Pull request by gvolpe79**

### CODE REVIEW

File: Detectors/GlobalTracking/src/MatchHMP.cxx [hunk 1]
```diff
@@ -258,7 +259,24 @@ void MatchHMP::addConstrainedSeed(o2::track::TrackParCov& trc, o2::dataformats::
...
+  float bxyz[3];
+  prop->getFieldXYZ(trc.getXYZGlo(), bxyz);
+  double bz = -bxyz[2];
+
+  float pCut = 0.;
+
+  if (TMath::Abs(bz - 5.0) < 0.5)
+    pCut = 0.450;
+  if (TMath::Abs(bz - 2.0) < 0.5)
+    pCut = 0.200;
```
Comment (shahor02) [inline Detectors/GlobalTracking/src/MatchHMP.cxx]: pleas use `{}`

File: Detectors/GlobalTracking/src/MatchHMP.cxx [hunk 2]
```diff
@@ -258,7 +259,24 @@ void MatchHMP::addConstrainedSeed(o2::track::TrackParCov& trc, o2::dataformats::
...
+
+  if (trc.getP() > pCut &&
+      TMath::Abs(trc.getEta()) < 0.52 &&
+      (TMath::RadToDeg() * trc.getPhi() < 90. || TMath::RadToDeg() * trc.getPhi() > 270.)) {
```
Comment (shahor02) [inline Detectors/GlobalTracking/src/MatchHMP.cxx]: please use 
```
if (trc.getP() > pCut && TMath::Abs(trc.getTgl()) < 0.544 && TMath::Abs(trc.getPhi() -TMath::Pi())<(TMath::Pi()*0.5)) { ...
```
to avoid unnecessary calculations.

File: Detectors/GlobalTracking/src/MatchHMP.cxx [hunk 3]
```diff
@@ -258,7 +259,24 @@ void MatchHMP::addConstrainedSeed(o2::track::TrackParCov& trc, o2::dataformats::
     mTracksLblWork[o2::globaltracking::MatchHMP::trackType::CONSTR].emplace_back(mRecoCont->getTPCITSTrackMCLabel(srcGID));
   }
 
-  mTracksIndexCache[o2::globaltracking::MatchHMP::trackType::CONSTR].push_back(it);
```
Comment (shahor02) [inline Detectors/GlobalTracking/src/MatchHMP.cxx:261]: Why do you need to cache the track into the mTracksWorkmTracksWork and mTrackGid and only then to apply fiducial selection? Why not caching only the tracks which can be used?

File: Detectors/GlobalTracking/src/MatchHMP.cxx [hunk 4]
```diff
@@ -377,14 +398,20 @@ void MatchHMP::doMatching()
...
+      // float maxTrkTime = (trackWork.second.getTimeStamp() + mSigmaTimeCut * timeUncert) * 1.E3; // maximum track time in ns
 
-      if (evtTime < (maxTrkTime + timeFromTF) && evtTime > (minTrkTime + timeFromTF)) {
+      float minTrkTime = (trackWork.second.getTimeStamp() - mSigmaTimeCut * timeUncert);
+      float maxTrkTime = (trackWork.second.getTimeStamp() + mSigmaTimeCut * timeUncert);
+
+      float deltaTime = evtTime - trackWork.second.getTimeStamp();
+
+      // if (evtTime < (maxTrkTime + timeFromTF) && evtTime > (minTrkTime + timeFromTF)) {
+      if (evtTime < maxTrkTime && evtTime > minTrkTime) {
```
Comment (shahor02) [inline Detectors/GlobalTracking/src/MatchHMP.cxx]: deltaTime is not used, either use it as 
```
float deltaTime = evtTime - trackWork.second.getTimeStamp();
if (std::abs(deltaTime) < mSigmaTimeCut * timeUncert) { 
...
```
and suppress minTrkTime and maxTrkTime or vice versa.

File: Detectors/GlobalTracking/src/MatchHMP.cxx [hunk 5]
```diff
@@ -458,11 +493,12 @@ void MatchHMP::doMatching()
           hmpTrk = nullptr;
           delete hmpTrkConstrained;
           hmpTrkConstrained = nullptr;
+          delete matching;
+          matching = nullptr;
+          oneEventClusters.clear();
           continue;
```
Comment (shahor02) [inline Detectors/GlobalTracking/src/MatchHMP.cxx:483]: There is no reason to use new/delete in these loops, instead of using `new` at 
https://github.com/AliceO2Group/AliceO2/blob/5f137b7119bc80e76ce18ef4d037535c6bc8e90a/Detectors/GlobalTracking/src/MatchHMP.cxx#L418
https://github.com/AliceO2Group/AliceO2/blob/5f137b7119bc80e76ce18ef4d037535c6bc8e90a/Detectors/GlobalTracking/src/MatchHMP.cxx#L426
https://github.com/AliceO2Group/AliceO2/blob/5f137b7119bc80e76ce18ef4d037535c6bc8e90a/Detectors/GlobalTracking/src/MatchHMP.cxx#L534 
and following deletes, just use stack objects (no need to delete them)

**Review by shahor02 — CHANGES_REQUESTED** [review]

File: Detectors/GlobalTracking/src/MatchHMP.cxx (hunk 1 above)
Comment (gvolpe79) [reply Detectors/GlobalTracking/src/MatchHMP.cxx]: Hi Ruben, thanks for your suggestions!
Sure I will add parenthesis!

File: Detectors/GlobalTracking/src/MatchHMP.cxx (hunk 2 above)
Comment (gvolpe79) [reply Detectors/GlobalTracking/src/MatchHMP.cxx]: ok, I will do it, but what getTgl() is? Why do you not like getEta()?

File: Detectors/GlobalTracking/src/MatchHMP.cxx (hunk 3 above)
Comment (gvolpe79) [reply Detectors/GlobalTracking/src/MatchHMP.cxx:261]: Yes, you are right, I will do like that!

File: Detectors/GlobalTracking/src/MatchHMP.cxx (hunk 4 above)
Comment (gvolpe79) [reply Detectors/GlobalTracking/src/MatchHMP.cxx]: deltaTime was for testing, then I forgot to cancel!

File: Detectors/GlobalTracking/src/MatchHMP.cxx (hunk 2 above)
Comment (shahor02) [reply Detectors/GlobalTracking/src/MatchHMP.cxx]: getTgl is a getter for the tan(lambda) (or 1/tan(theta)) : on of the track parameters, so it comes for no cost.
getEta() is calculated from it via quite expensive calls: https://github.com/AliceO2Group/AliceO2/blob/5b4a79edcad26f04d1052ba8093155e327c5c3fc/DataFormats/Reconstruction/include/ReconstructionDataFormats/TrackParametrization.h#L597-L608

File: Detectors/GlobalTracking/src/MatchHMP.cxx (hunk 5 above)
Comment (gvolpe79) [reply Detectors/GlobalTracking/src/MatchHMP.cxx:483]: so for "matching" was my first choice but then since I pass it to a method  of an other class that modifies it, does not work, I need to use the pointer! If I do just like this "MatchInfo *matching" it compiles but crashes in execution.
Then I need to create copies of the tracks (trackHMP and hmpTrkConstrained), can I do without using "= new ...."?

Comment (shahor02) [reply Detectors/GlobalTracking/src/MatchHMP.cxx:483]: For the `matching`: you can simply use at https://github.com/AliceO2Group/AliceO2/blob/5f137b7119bc80e76ce18ef4d037535c6bc8e90a/Detectors/GlobalTracking/src/MatchHMP.cxx#L418
```
MatchInfo (999999, mTrackGid[type][cacheTrk[itrk]]);
```
then change all `matching->` by `matching.`, all `*matching` by `matching` and passing the pointer to `&matching`.
And remove all the lines like https://github.com/AliceO2Group/AliceO2/blob/5f137b7119bc80e76ce18ef4d037535c6bc8e90a/Detectors/GlobalTracking/src/MatchHMP.cxx#L437-L438

Idem for https://github.com/AliceO2Group/AliceO2/blob/5f137b7119bc80e76ce18ef4d037535c6bc8e90a/Detectors/GlobalTracking/src/MatchHMP.cxx#L426 and https://github.com/AliceO2Group/AliceO2/blob/5f137b7119bc80e76ce18ef4d037535c6bc8e90a/Detectors/GlobalTracking/src/MatchHMP.cxx#L534 (and you don't need to define the `hmpTrkConstrained` pointer in advance).

File: Detectors/GlobalTracking/include/GlobalTracking/MatchHMP.h [hunk 6]
```diff
@@ -109,9 +109,9 @@ class MatchHMP
                             ITSTPCTRD,
                             SIZEALL };
 
-  std::vector<o2::dataformats::MatchInfoHMP>& getMatchedTrackVector(o2::globaltracking::MatchHMP::trackType index) { return mMatchedTracks[index]; }
+  std::vector<o2::dataformats::MatchInfoHMP> getMatchedTrackVector(o2::globaltracking::MatchHMP::trackType index) { return mMatchedTracks[index]; }
 
-  std::vector<o2::MCCompLabel>& getMatchedHMPLabelsVector(o2::globaltracking::MatchHMP::trackType index) { return mOutHMPLabels[index]; } ///< get vector of HMP label of matched tracks
+  std::vector<o2::MCCompLabel> getMatchedHMPLabelsVector(o2::globaltracking::MatchHMP::trackType index) { return mOutHMPLabels[index]; } ///< get vector of HMP label of matched tracks
```
Comment (shahor02) [inline Detectors/GlobalTracking/include/GlobalTracking/MatchHMP.h]: Why did you change the returned type from the reference to copy?

File: Detectors/GlobalTracking/src/MatchHMP.cxx (hunk 3 above)
Comment (shahor02) [reply Detectors/GlobalTracking/src/MatchHMP.cxx:261]: still don't understand why do you need to push to `mTracksIndexCache` if the track is not used. Now you are adding the same `it` multiple times and then loop over these multiple entries. I think mTracksIndexCache also should be updated only when the track is accepted.

File: Detectors/GlobalTracking/src/MatchHMP.cxx [hunk 7]
```diff
@@ -377,41 +398,39 @@ void MatchHMP::doMatching()
...
-        matching->setIdxTrack(trackGid);
+        matching.setHMPIDtrk(0, 0, 0, 0);            // no intersection found
+        matching.setHMPIDmip(0, 0, 0, 0);            // store mip info in any case
+        matching.setIdxHMPClus(99, 99999);           // chamber not found, mip not yet considered
+        matching.setHMPsignal(Recon::kNotPerformed); // ring reconstruction not yet performed
+        matching.setIdxTrack(trackGid);
 
-        TrackHMP* hmpTrk = new TrackHMP(trefTrk); // create a hmpid track to be used for propagation and matching
-        TrackHMP* hmpTrkConstrained = nullptr;    // create a hmpid track to be used for propagation and matching
+        TrackHMP hmpTrk = TrackHMP(trefTrk); // create a hmpid track to be used for propagation and matching
```
Comment (shahor02) [inline Detectors/GlobalTracking/src/MatchHMP.cxx]: use simply ` TrackHMP hmpTrk(trefTrk);` to avoid extra assignment.

File: Detectors/GlobalTracking/src/MatchHMP.cxx [hunk 8]
```diff
@@ -492,39 +509,32 @@ void MatchHMP::doMatching()
...
         // 4. Propagate back the constrained track to the radiator radius
 
-        hmpTrkConstrained = new TrackHMP(trackC);
-        hmpTrkConstrained->set(trackC.getX(), trackC.getAlpha(), trackC.getParams(), trackC.getCharge(), trackC.getPID());
-        if (!prop->PropagateToXBxByBz(*hmpTrkConstrained, radiusH - kdRadiator, o2::base::Propagator::MAX_SIN_PHI, o2::base::Propagator::MAX_STEP, matCorr)) {
-          delete hmpTrk;
-          hmpTrk = nullptr;
-          delete hmpTrkConstrained;
-          hmpTrkConstrained = nullptr;
+        TrackHMP hmpTrkConstrained = TrackHMP(trackC);
```
Comment (shahor02) [inline Detectors/GlobalTracking/src/MatchHMP.cxx]: idem, use `TrackHMP hmpTrkConstrained(trackC);`

File: Detectors/GlobalTracking/src/MatchHMP.cxx [hunk 9]
```diff
@@ -377,41 +398,39 @@ void MatchHMP::doMatching()
...
+      float maxTrkTime = (trackWork.second.getTimeStamp() + mSigmaTimeCut * timeUncert);
 
-      if (evtTime < (maxTrkTime + timeFromTF) && evtTime > (minTrkTime + timeFromTF)) {
+      // if (evtTime < (maxTrkTime + timeFromTF) && evtTime > (minTrkTime + timeFromTF)) {
+      if (evtTime < maxTrkTime && evtTime > minTrkTime) {
 
         evtTracks++;
 
-        MatchInfo* matching = new o2::dataformats::MatchInfoHMP(999999, mTrackGid[type][cacheTrk[itrk]]);
+        MatchInfo matching = MatchInfo(999999, mTrackGid[type][cacheTrk[itrk]]);
```
Comment (shahor02) [inline Detectors/GlobalTracking/src/MatchHMP.cxx]: use `MatchInfo matching(999999, mTrackGid[type][cacheTrk[itrk]]);` to avoid extra assignment.

**Review by shahor02 — CHANGES_REQUESTED** [review]

File: Detectors/GlobalTracking/include/GlobalTracking/MatchHMP.h (hunk 6 above)
Comment (gvolpe79) [reply Detectors/GlobalTracking/include/GlobalTracking/MatchHMP.h]: Hi Ruben, because when I changed the MatchInfo pointer object to stack object as you suggested, the wokflow failed and like this not!

File: Detectors/GlobalTracking/src/MatchHMP.cxx (hunk 7 above)
Comment (gvolpe79) [reply Detectors/GlobalTracking/src/MatchHMP.cxx]: Sure, of course it is wrong! By mistake the push it is outside the if() ...! IThanks to have spotted it!

File: Detectors/GlobalTracking/include/GlobalTracking/MatchHMP.h (hunk 6 above)
Comment (shahor02) [reply Detectors/GlobalTracking/include/GlobalTracking/MatchHMP.h]: getMatchedTrackVector is used only to pass the vector to DPL snapchot, I don't see how it can be related to changing the pointer by object well before that. What was the failure?

Comment (gvolpe79) [reply Detectors/GlobalTracking/include/GlobalTracking/MatchHMP.h]: It is was strange also for me! Anyway, now I retried again, and it is working, maybe there was an other reason! I have just committed the revised code.
